### DISCUSSION TITLE
How to build for Apple M1?

### DISCUSSION THREAD

**Discussion by dxlr8r** [FreeTubeApp/FreeTube#2771] [Q&A] [answered]
Have read through:

https://github.com/FreeTubeApp/FreeTube/issues/2054
https://github.com/FreeTubeApp/FreeTube/pull/1918

And from what I can understand:

```
git clone https://github.com/FreeTubeApp/FreeTube.git
cd FreeTube
git checkout development
yarn install
npm run build
```

That however only build x86 artifacts.

**Comment by absidue** [comment, marked as answer]
instead of `npm run build` you want to do `yarn run build:arm64`

**Reply by dxlr8r** [reply]
Thank you! Looking forward to official support. For now happy with my own build.

**Reply by absidue** [reply]
Unfortunately macOS gatekeeper, blocks M1 builds unless they are signed with a developer or enterprise certificate. It doesn't complain if you run unsigned builds if they were built on your own computer. It seems like using the x86 builds through Rosetta, stops gatekeeper from gatekeeping.
See https://github.com/FreeTubeApp/FreeTube/pull/2113#issuecomment-1163950905 for more details.

**Reply by dxlr8r** [reply]
Couldn’t you add a Gatekeeper exception,`sudo spctl --add`?

If not, instead of binary, add it to brew or MacPorts as a source package?

**Reply by PikachuEXE** [reply]
Since the x64 version works mostly I think we will be releasing that version for now for Mac
Unless we can find/are provided some example(s) of producing arm64 version without cert

**Reply by dxlr8r** [reply]
In order to use the x64 version you have to install Rosetta 2, and even then there is additional cost in running the software. Rosetta is awesome, but a native binary is always better.

x64 is dead for Apple, and like Rosetta 1 (PPC->x86) version 2 has a limited life span. One can kick the can down the road for some time, but there is no going back.

**Reply by absidue** [reply]
@dxlr8r Asking users to add a gatekeeper exception is a bad user experience. We don't maintain the homebrew or MacPorts packages, if you would like to use source packages, please ask their respective maintainers to add them.

**Reply by lothar-cell** [reply]
> x64 is dead for Apple, and like Rosetta 1 (PPC->x86) version 2 has a limited life span. One can kick the can down the road for some time, but there is no going back.

It's true enough, that the sooner the better.

**Reply by Brog33** [reply]
> Unless we can find/are provided some example(s) of producing arm64 version without cert

any progress on this?

**Reply by 12people** [reply]
@PikachuEXE 0 A.D. works that way, and they simply ask the user to run `xattr -cr /Applications/0\ A.D..app` on the [download page](https://play0ad.com/download/mac/).

@absidue 
I know that isn't the best UX, but the Intel version is also not a great UX, with slow startup times and being resource-heavy in general.

Perhaps present the ARM version as only for "advanced" users who are OK running this cert command? Or perhaps only offer it here as a Github package, without an official download link on the website?

Either way, it'd be great if there was a way for ARM Mac users to run an official native version on their device.

**Reply by robigan** [reply]
> @dxlr8r Asking users to add a gatekeeper exception is a bad user experience. We don't maintain the homebrew or MacPorts packages, if you would like to use source packages, please ask their respective maintainers to add them.

How so? It's the easiest thing to do and many apps in Homebrew are not signed, all you have to do is right click on the app and click open from Finder, this bypasses Gatekeeper very easily, without any terminal. I've been doing this for years and never had an issue, personally I don't see why Gatekeeper should be an issue regardless of signing or not.

**Reply by 12people** [reply]
@robigan 
Actually, on ARM-based Macs, it's not as simple (the right-click trick doesn't work) and one has to use the Terminal.

That said, I think x86 FreeTube on ARM Macs is a bad user experience too and using Homebrew itself requires Terminal usage, so I don't really see much of an argument why not have an ARM build on Homebrew.

**Reply by robigan** [reply]
> @robigan 
> 
> Actually, on ARM-based Macs, it's not as simple (the right-click trick doesn't work) and one has to use the Terminal.
> 
> 
> 
> That said, I think x86 FreeTube on ARM Macs is a bad user experience too and using Homebrew itself requires Terminal usage, so I don't really see much of an argument why not have an ARM build on

Pretty sure I just did the right click trick on a M1 Pro 30 minutes ago cus Gatekeeper was complaining about FreeTube...

**Reply by absidue** [reply]
@12people The FreeTube cask in the main/default Homebrew  tap is not maintained by the FreeTube team (which is also why it is listed in the unofficial downloads section in the README), so if you want an ARM64 version there, you need to ask whoever maintains the main/default Homebrew tap.

That being said you can use this unofficial Homebrew tap that has an ARM64 version: https://github.com/PikachuEXE/homebrew-FreeTube

**Reply by robigan** [reply]
> @12people The FreeTube cask in the main/default Homebrew  tap is not maintained by the FreeTube team (which is also why it is listed in the unofficial downloads section in the README), so if you want an ARM64 version there, you need to ask whoever maintains the main/default Homebrew tap.
> 
> That being said you can use this unofficial Homebrew tap that has an ARM64 version: https://github.com/PikachuEXE/homebrew-FreeTube

Was gonna do the ARM64 myself actually. I'll try to make a PR into casks and see how that goes.

**Reply by robigan** [reply]
Ok, so turns out I made some mistakes. It is true that you can bypass Gatekeeper using right-click, [but according to Apple's own docs](https://support.apple.com/en-us/102445), when it displays a prompt such as "App is damaged", it seems to be because of some security detection triggering for arm64 builds. On Intel builds, it could be that either Gatekeeper is disabling itself due to Rosetta 2 requirement, or it is properly validating Intel builds for malware and just complaining of missing notarization, but I don't think it's the former as Gatekeeper works on Intel, so why can't the same checks be performed on Intel-made software regardless of CPU architecture? (I mean, Gatekeeper from what I've understood just checks for certain hashes of malicious or safe software, unless I'm missing something). 



> > @12people The FreeTube cask in the main/default Homebrew  tap is not maintained by the FreeTube team (which is also why it is listed in the unofficial downloads section in the README), so if you want an ARM64 version there, you need to ask whoever maintains the main/default Homebrew tap.
> > That being said you can use this unofficial Homebrew tap that has an ARM64 version: https://github.com/PikachuEXE/homebrew-FreeTube
> 
> Was gonna do the ARM64 myself actually. I'll try to make a PR into casks and see how that goes.

What I meant by this was that I thought Homebrew had the option of building from source for casks, but building from source is a thing only for formulae as I discovered. Essentially, binaries are only allowed in casks. I think it would be nice if the build system can build for Arm64, and then reference that in the Homebrew casks. Not asking at all to add arm64 builds to the FreeTube website... Although if the issue of Gatekeeper complaining about damaged software is really simply just what I said above, then perhaps that issue could be fixed.

**Reply by robigan** [reply]
Did some more research, although effectively useless in terms of security, the bundles must be Ad-Hoc signed, and cannot be plainly unsigned.

As seen here, using the bundle by @PikachuEXE 
<img width="1512" alt="image" src="https://github.com/FreeTubeApp/FreeTube/assets/35210888/6099e8f4-e2e3-4197-85d5-2a6821b5566c">

See (in order of how I dug my rabbit hole) https://github.com/nwjs/nw.js/issues/8157, https://apple.stackexchange.com/questions/288291/what-are-the-restrictions-of-ad-hoc-code-signing, https://wiki.lazarus.freepascal.org/Code_Signing_for_macOS#Ad_hoc_signing, and https://gist.github.com/rsms/929c9c2fec231f0cf843a1a746a416f5

**Reply by robigan** [reply]
P.S. I'd be more than happy to make PRs

**Reply by chasewalsh** [reply]
Thanks.  It works perfectly and I'm kicking myself why I didn't do this earlier.  The x86 version is very buggy and the Arm64 version is butter smooth after compiling it with these instructions.

**Comment by ngocphamm** [comment]
Not sure if anyone else is interested, but if you have Bun installed, you can also do `bun install` and `bun run build:arm64`. Takes only a few minutes on my M1 Max.

**Comment by Dodoooh** [comment]
Hi, this script automates the setup and ARM64 build for FreeTube on macOS. It clones the repository (if needed), updates it, installs dependencies, and builds the app. After a setup, only ./build_freetube.sh can be executed and the new app can be moved from folder /build to the applications 

Usage:

	1.	Save as build_freetube.sh
	2.      Make sure to set the `REPO_DIR` variable to the correct path.
	3.	Make executable: chmod +x build_freetube.sh
	4.	Run: ./build_freetube.sh

```
#!/bin/bash

# FreeTube Build Script for macOS
# Created by [Dodoooh]
# This script automates the setup and ARM64 build process for FreeTube.
# Make sure to set the `REPO_DIR` variable to the correct path.

# Set the directory for the FreeTube repository
REPO_DIR="/Users/username/projects/FreeTube"

# Check if required tools are installed
for cmd in git npm yarn; do
  if ! command -v $cmd &> /dev/null; then
    echo "Error: $cmd is not installed. Please install it and try again."
    exit 1
  fi
done

# Ask for confirmation before starting the installation
read -p "Do you want to start the FreeTube build process? (y/n): " confirm
if [[ $confirm != [yY] ]]; then
  echo "Build process canceled."
  exit
fi

# Step 1: Clone the repository if it doesn't already exist
if [ ! -d "$REPO_DIR" ]; then
  echo "Cloning FreeTube repository..."
  git clone https://github.com/FreeTubeApp/FreeTube.git "$REPO_DIR"
fi

# Step 2: Navigate to the repository directory
cd "$REPO_DIR" || exit

# Step 3: Switch to the development branch and pull the latest changes
echo "Switching to the development branch..."
git checkout development
echo "Pulling the latest changes..."
git pull

# Step 4: Install dependencies with npm and yarn
echo "Installing dependencies with npm..."
npm install
echo "Installing dependencies with yarn..."
yarn install

# Step 5: Build for ARM64
echo "Building for ARM64..."
yarn run build:arm64

echo "Build complete!"
```
